chore: Introduce Kamelet input/output data types

christophd opened the pull request:
- Introduce data type converters
- Add data type processor to auto convert exchange message from/to given data type
- Let user choose which data type to use (via Kamelet property)
- Add data type registry and annotation based loader to find data type implementations by component scheme and name

Relates to CAMEL-18698 and https://github.com/apache/camel-k/issues/1980

This PR introduces input/output data types on Kamelets. Each Kamelet is able to use a specific data type processor and a registry to resolve data types and its conversion logic.

```yaml
apiVersion: camel.apache.org/v1alpha1
kind: Kamelet
metadata:
  name: aws-s3-source
  labels:
    camel.apache.org/kamelet.type: "source"
spec:
  template:
    beans:
    - name: dataTypeRegistry
       type: "#class:org.apache.camel.kamelets.utils.format.DefaultDataTypeRegistry"
    - name: dataTypeProcessor
      type: "#class:org.apache.camel.kamelets.utils.format.DataTypeProcessor"
      property:
        - key: format
          value: '{{outputFormat}}'
[...]
      steps:
      - process:
          ref: "{{dataTypeProcessor}}"
      - to: "kamelet:sink"
```

The user chooses the data type in the Kamelet binding by its format name and sets the property `outputFormat` on the binding source.

```yaml
apiVersion: camel.apache.org/v1alpha1
kind: KameletBinding
metadata:
  name: aws-s3-uri-binding
spec:
  source:
    ref:
      kind: Kamelet
      apiVersion: camel.apache.org/v1alpha1
      name: aws-s3-source
    properties:
      bucketNameOrArn: myBucket
      accessKey: ...
      secretKey: ...
      region: …
      outputFormat: json
  sink:
    uri: log:info
```

The `json` data type is provided by the aws-s3 component and adds automatic message body conversion logic. The data type and its conversion logic is defined via annotations and a registry automatically performs a lookup based on the component scheme and the data format name.

As an example the PR introduces some standard data type converters as well as output data types for the `aws-s3-source` and input data types for the `aws-ddb-sink` Kamelet.

lburgazzoli commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/converter/aws2/s3/AWS2S3JsonOutputType.java:
```diff
@@ -0,0 +1,63 @@
+public class AWS2S3JsonOutputType implements DataTypeConverter {
+
+    private static final String TEMPLATE = "{" +
+            "\"key\": \"%s\", " +
```
I'm not sure to understand the use case for this converter as it would mean that the body would need to be unwrapped every time to access the actual content and what happens if the actual payload is already a json or another maybe binary format ?

christophd replied:
yes, true! this is not rock solid. I was playing with some output format and came up with this. 

But I think data types should be opinionated (same as Kamelets) covering a specific set of use cases. Like using this json format with binary file content just won't work as you have already mentioned. 

The idea was to combine specific header information like AWSS3Key and the file content into a Json structure that follow-up steps/components can easily work with. The output format can also expose a Json schema so others know what to expect in the Json structure.

christophd replied:
@lburgazzoli I have added a CloudEvent output data type for AWS S3. This might make more sense as an example. WDYT?

christophd replied:
Basically the output data type sets CloudEvent specific headers with AWS S3 specific information. This way the user is able to directly map this to Knative (Kamelet binding aws-s3-to-knative) for instance.

christophd replied:
see library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/converter/aws2/s3/AWS2S3CloudEventOutputType.java

lburgazzoli commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/converter/standard/JsonModelDataType.java:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        String type = exchange.getProperty(JSON_DATA_TYPE_KEY, String.class);
+        try (JacksonDataFormat dataFormat = new JacksonDataFormat(new ObjectMapper(), Class.forName(type))) {
```
Creating an ObjectMapper per invocation seems quite expensive.

Ideally an object mapper from the registry should be used if present (i.e. you may want to tweak how the object mapper works), if not a local but cached one can be used.

christophd replied:
Done

lburgazzoli commented on kamelets/aws-ddb-sink.kamelet.yaml:
```diff
@@ -107,17 +113,24 @@ spec:
   - "camel:kamelet"
   template:
+    beans:
+    - name: dataTypeRegistry
```
Not sure if this should be there as there will be an instance of the registry for each kamelet instantiation

christophd replied:
Good point. How can I make sure to add this to the Camel registry as a singleton service?

lburgazzoli replied:
it must be done outside kamelets but eventually you can explore about not needing a registry at all ad lazy loading converters.

christophd replied:
see https://github.com/apache/camel-k/issues/3845

lburgazzoli commented on kamelets/aws-s3-source.kamelet.yaml:
```diff
@@ -107,13 +107,28 @@ spec:
         description: The number of milliseconds before the next poll of the selected bucket.
         type: integer
         default: 500
+      outputFormat:
```
the list of possible  types must be expressed through an enum

christophd replied:
why limiting this to a predefined enum of formats? The user may provide a custom format, too.

lburgazzoli replied:
the problem is: how the user/tooling know what converters are available and what are supported ?

christophd replied:
its in the Kamelet specification. The default supported output/input types are specified and described in the Kamelet spec. Like this:

```yaml
apiVersion: camel.apache.org/v1alpha1
kind: Kamelet
metadata:
  name: aws-s3-source
  labels:
    camel.apache.org/kamelet.type: "source"
spec:
  definition:
    title: "AWS S3 Source"
    type: object
    properties:
      bucketNameOrArn:
        title: Bucket Name
        description: The S3 Bucket name or ARN
        type: string
      ...
  output:
    default: binary
    types:
      binary:
        mediaType: application/octet-stream
      json:
        mediaType: application/json
        dependencies:
          - "camel:jackson"
        schema:
          type: object
          properties:
            key:
              title: S3 key
              description: The S3 key identifier
              type: string
            fileContent:
              title: File content
              description: The file content as String
              type: string
```

This is the place where each data type is able to specify a schema and additional dependencies, too.

lburgazzoli replied:
yes but if we expect the user to provide custom converter, then  the only way to validate it is to wait for runtime failures. I'm thinking about how tooling would provide support/validation for that.

lburgazzoli commented on kamelets/aws-ddb-sink.kamelet.yaml:
```diff
@@ -97,6 +97,12 @@ spec:
         x-descriptors:
           - 'urn:alm:descriptor:com.tectonic.ui:checkbox'
         default: false
+      inputFormat:
```
the list of possible types must be expressed through an enum

christophd replied:
same as for output formats. I think the user may provide a custom format, too.

oscerd replied:
yes, but at least as description or as information, we need a list of default types.

christophd replied:
@oscerd valid concern and I think we can add a list of default types. this is to be done in a separate PR though once Camel K Kamelet CRDs are updated with additional data type fields.

lburgazzoli commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/DataTypeProcessor.java:
```diff
@@ -0,0 +1,67 @@
+            return;
+        }
+
+        dataTypeRegistry.lookup(scheme, format)
```
the lookup can likely be performed only once

christophd replied:
yes, true the format is not going to change for the KameletBinding instance so we can cache the converter once the lookup das been performed. thanks!

christophd replied:
Caching the converter in the processor once lookup has been performed

lburgazzoli reviewed (commented):
Added some initial comments

lburgazzoli commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/DefaultDataTypeRegistry.java:
```diff
@@ -0,0 +1,154 @@
+ *
+ * The registry is able to retrieve converters for a given data type based on the component scheme and the given data type name.
+ */
+public class DefaultDataTypeRegistry extends ServiceSupport implements DataTypeRegistry, CamelContextAware {
```
I wonder if we need a registry or we could lazy load the converter when the processor is created and initialized using the standard camel factory finder mechanism which would work out of the box  also on quarkus

christophd replied:
Cool! I will take a look on the factory finder mechanism. The idea with the registry was to have this ready for a broader scope (not only for Kamelets but in Camel core). For the Kamelet use case in particular this might be a heavyweight solution, agreed.

christophd replied:
I have added a lazy loading of component converters via resource path lookup using the factory finder mechanism

christophd replied:
we need to enable the factory finder in camel-quarkus for this with `quarkus.camel.service.discovery.include-patterns` build time property. This needs to be done in Camel K so I will add a new issue to track this one in Camel K.

christophd replied:
see https://github.com/apache/camel-k/issues/3844

lburgazzoli replied:
Another option would be to create a quarkus extension which would register the service and i.e. also the factory finder without the need to have properties

oscerd reviewed (changes requested):
I just have some worries about leaving the types without a default types list.

claudio4j commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/AnnotationDataTypeLoader.java:
```diff
@@ -0,0 +1,171 @@
+/**
+ * Data type loader scans packages for {@link DataTypeConverter} classes annotated with {@link DataType} annotation.
+ */
+public class AnnotationDataTypeLoader implements DataTypeLoader, CamelContextAware {
```
I wonder, if this annotation and loader is going to work when running in native mode ?

christophd replied:
good pint. not sure either about this. we will switch to factory finder mechanism once this is enabled for the DataTypeConverter in camel-quarkus and this mechanism will work in native mode, too.

christophd commented:
@oscerd I was adding some more commits while iterating and improving the PR. Do you want me to squash the shanges to some essential commits first?

christophd commented:
also I see I need to rebase

oscerd commented:
That's fine, no need for squash at least from my PoV

oscerd commented:
Let me know once is ready and I'll review and merge. We need also to identify the default type for each of Kamelets we're going to modify.

oscerd commented:
Camel-k-runtime seems to be ok while upgrading to the required bits for Camel 3.19.0 and Camel Quarkus 2.14.0. I would like to include this PR in 0.10.0 and release by the end of this week or beginning of the next. So we could have a first camel k release (1.11.0) upstream with this feature.

oscerd commented:
Can you rebase @christophd ? Thanks.

davsclaus commented:
Please make sure that this works in standalone Camel as well, eg such as try via camel-jbang.
Kamelets are universal building blocks and MUST work on camel-spring-boot, camel-kafka-connector, camel-quarkus (without camel-k), and of course camel-k as well.

Previously Kamelets have been kept more basic where the kamelet-util JAR was what the name said - utility. With this kind of work, then its no longer util, but a core and api part of kamelets, which IMHO later should be moved into their own modules.

oscerd commented:
We could create a kamelet-api module. I'll try the PR on jbang too.

christophd commented:
@davsclaus absolutely! IMO this should go to Camel core in the long term. That's why I was adding all those spi and api parts as well as the registry with factory finder lookup as part of this solution.

christophd commented:
@oscerd all checks have passed now. I think this is ready for a final review. 

I have provided some fixes to YAKS tests in general so they are all successful now (before the tests actually have been skipped reporting green jobs). I will backport the YAKS related fixes to 0.9 once this is merged.

oscerd commented on library/camel-kamelets-utils/pom.xml:
```diff
@@ -71,12 +71,22 @@
+        <dependency>
+            <groupId>org.apache.camel</groupId>
+            <artifactId>camel-cloudevents</artifactId>
+            <!--<scope>provided</scope>-->
```
Do we need to remove this?

christophd replied:
It is used in AWS S3 CloudEvents output data type.

The thing is `camel-cloudevents` is not available in camel-quarkus. Because of this the dependency is not able to resolve via CamelCatalog at runtime because the catalog only knows camel-quarkus extensions. When adding it here as a dependency it will be resolved as a transitive dependency using the Camel core version.

We should add it to camel-quarkus and use scope provided here so Camel K is able to resolve this dependency for bindings that actually use the data type.

christophd replied:
I had another look and the AWS S3 data type only uses some constants related to header names in camel-cloudevents. I will use plain String header names and remove the dependency for now.

christophd replied:
I removed the dependency in favor of using plain String constants

lburgazzoli replied:
there is a camel-cloudevents extension https://github.com/apache/camel-quarkus/tree/main/extensions/cloudevents

oscerd reviewed (approved):
Looks good to me. Would like to have more feedback from @lburgazzoli and @davsclaus 

We need to ensure this stuff will work across all the runtimes.

davsclaus commented on library/camel-kamelets-utils/src/main/resources/META-INF/services/org/apache/camel/DataTypeConverter:
```diff
@@ -0,0 +1,20 @@
+# limitations under the License.
+#
+
+org.apache.camel.kamelets.utils.format.converter.standard
```
In the core camel project we source generate these META-INF files. We should do this in the future for kamelets too

christophd commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/converter/aws2/s3/AWS2S3BinaryOutputType.java:
```diff
@@ -0,0 +1,55 @@
+        try {
+            InputStream is = exchange.getMessage().getBody(InputStream.class);
+            if (is != null) {
+                exchange.getMessage().setBody(IoUtils.toByteArray(is));
```
good finding, Claus! It used to have AWS specific ResponseInputStream handling before but the I did refactor it to use generic InputStream so we can now have this as a plain camel data type converter. thx

christophd replied:
Done

christophd commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/spi/DataTypeRegistry.java:
```diff
@@ -0,0 +1,60 @@
+     * @param converter
+     */
+    default void addDataTypeConverter(DataTypeConverter converter) {
+        addDataTypeConverter("camel", converter);
```
the `camel` scheme represents the default data type converters. These are present in factory finder with `camel-*` and get used as a fallback when registry is not able to find a specific converter for the component scheme.

christophd commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/spi/DataTypeConverter.java:
```diff
@@ -0,0 +1,63 @@
+import org.apache.camel.kamelets.utils.format.spi.annotations.DataType;
+
+@FunctionalInterface
+public interface DataTypeConverter {
```
👍 Done

christophd commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/spi/DataTypeConverterResolver.java:
```diff
@@ -0,0 +1,49 @@
+     * @param camelContext
+     * @return
+     */
+    Optional<DataTypeConverter> resolve(String scheme, String name, CamelContext camelContext);
```
👍 Done

lburgazzoli commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/converter/standard/JsonModelDataType.java:
```diff
@@ -0,0 +1,85 @@
+        ObjectHelper.notNull(camelContext, "camelContext");
+
+        String type = exchange.getProperty(DATA_TYPE_MODEL_PROPERTY, String.class);
+        try (JacksonDataFormat dataFormat = new JacksonDataFormat(mapper, camelContext.getClassResolver().resolveMandatoryClass(type))) {
```
This seems quite heavy as you need to create a class from a string for each message
IMHO, as a followup there should be a way to configure the converter properties i.e. in this case I think the type is likely to be a fixed.

lburgazzoli replied:
Also note that the a data format it likely needed to be[ initialized/started ](https://github.com/apache/camel/blob/bf98522eb05c8e477c182f11d15c302e69d0bc0d/components/camel-jackson/src/main/java/org/apache/camel/component/jackson/AbstractJacksonDataFormat.java#L531)

christophd replied:
@oscerd @lburgazzoli @davsclaus I have simplified the JsonModel data type converter to not use the JackSonDataFormat but a normal ObjectMapper that is shared on a class instance level

oscerd commented:
I would like to release 0.10.0 this week. There are some leftover to work on @christophd I guess. I'll wait all green in terms of review to merge.

christophd commented on library/camel-kamelets-utils/src/main/java/org/apache/camel/kamelets/utils/format/converter/aws2/s3/AWS2S3CloudEventOutputType.java:
```diff
@@ -0,0 +1,58 @@
+    public void convert(Exchange exchange) {
+        final Map<String, Object> headers = exchange.getMessage().getHeaders();
+
+        headers.put(CAMEL_CLOUD_EVENT_TYPE, "kamelet.aws.s3.source");
```
Ah, these are good suggestions! So the type would be something like `org.apache.camel.event.aws.s3.{{operation}}` and the source something like `aws.s3.{{bucketName}}`

oscerd commented:
I'm still under the impression this PR is WIP. We are not sure this will work across all the runtimes, we have some missing bits for make it work on Quarkus and camel-k. Probably it would be better to postpone the merge to another kamelets release, when we have something more stable. Thoughts? @lburgazzoli @christophd @davsclaus

lburgazzoli commented:
> I'm still under the impression this PR is WIP. We are not sure this will work across all the runtimes, we have some missing bits for make it work on Quarkus and camel-k. Probably it would be better to postpone the merge to another kamelets release, when we have something more stable. Thoughts? @lburgazzoli @christophd @davsclaus

IMHO we must be sure that introducing this new feature does not end up break existing running integrations.
If there is a way to make this feature an opt-in, then we can merge it and move forward

christophd commented:
The changes should not harm existing runtimes and integrations. The operator may build a new integration kit because of the additional utils dependency but that should be all. 

The YAKS tests that we have should make sure that the DDB sink Kamelet is still working as it was before. I have not changed the tests and they run successful in this PR. Also we have new YAKS tests for AWS S3 source that also verify the default behavior of the Kamelet as it has been before.

@oscerd @lburgazzoli I think we can wait for the end of the day and then make a decision. I can work on some of the open points today so we should only have these left overs that we can tackle in follow-up iterations: 

- Make factory finder mechanism work in camel-quarkus (may lead to a new Camel K release)
- Disable classpath scan by default and use lazy loading via factory finder as a result of the point above
- Generate factory finder META-INF files as done in Camel core
- Add DataTypeRegistry as singleton to the Camel registry to avoid local instances per Kamelet

oscerd commented:
There is one point in the PR I didn't notice before, creating a Jackson Data format for each message is really overkilling. My only concern is related to the fact the behavior will be or not the same, in all the runtimes. We are introducing a new parameter with a default value, so the kamelet lifecycle and execution path won't be the same, I suppose.

oscerd commented:
Since, at least, from my PoV, this is a really big change, I would maybe postpone this to after 3.20.0 release.

davsclaus commented:
This PR changes 2 existing kamelets so it cannot be an opt-in. 
So I suggest to revert the 2 kamelets so all kamelets works as now.
And then add new kamelets that use this new stuff and mark them as experimental.

davsclaus commented:
This gives time for this new stuff to stabilize and improve and have it moved out of kamelets-utils as its not util and into its own module. And then later find out what can/should go into core camel.

And we need to ensure that this works across all Camel runtimes (CSB, CEQ, CKC, JBang, ...) which it ought to do as it uses camel practices, but we haven't had time in these other projects to test and try.

oscerd commented:
Yes, I think something like introducing some side kamelets with the new feature and test them would be best. Then we can introduce this slowly on the Kamelets, once we have everything in the correct place.

oscerd reviewed (changes requested):
LGTM, there is just one point I'm questioning:
Wasn't it better to create something like aws-s3-source-exp and aws-ddb-sink-exp in the kamelets directory instead of putting them in a subfolder? At least with the 1.11.0 release of camel-k and camel-kamelets we could have them listed as available and installed.

davsclaus commented:
camel-kamelet (kamelets) are discovered via camel-kamelets-catalog in the kamelets sub folder, so you should NOT have other folders.

If we end up with alot of kamelets, then we can work on supporting folder hierarchy instead of single folder as currently.

christophd commented:
@davsclaus yes, good to know. 

at the moment the experimental folder is not inside `kamelets` but right next to it as a top level folder. so experimental Kamelets will not be included in the catalog at the moment. 

@oscerd suggested to add those experimental Kamelets to the catalog. I am fine with it. When thinking about it it might be good to include them because then people can give it a try. I could add `@experimental` tag to the tests for these Kamelets so we can separate the tests, too.

oscerd commented:
@christophd if you rebase the validation should be fixed now.

christophd commented:
@oscerd unfortunately is is still failing after rebase

oscerd commented:
Pushed on main, should work now if you rebase.

christophd commented:
@oscerd @lburgazzoli @davsclaus look at this all checks have passed 😄 

So what we have right now is following:
- reverted AWS S3 source and AWS DDB sink Kamelet to not use data types and they work as it has been before
- introduced experimental Kamelets (in experimental top level folder) that make use of data types utils
- add some YAKS test coverage for experimental Kamelets that are also part of the GitHub actions CI jobs

Only question is if we should include the experimental Kamelets into the catalog right now or leave them in experimental folder for now. I am happy with both ways but I guess it will be easier for people to give it a try when they are part of the catalog.

oscerd commented:
Moving them to the official catalog seems to be better maybe something like aws-s3-source-exp.kamelet.yaml instead of .exp.kamelet.yaml. So we could give them to the users.

oscerd commented:
and have early feedback

oscerd commented:
So I would prefer to have them in the catalog from the 0.10.0 release.

christophd commented:
@oscerd the experimental Kamelets are now included in the catalog. Because of validation scripts I ended up with following naming:

- aws-s3-experimental-source
- aws-ddb-experimental-sink

All checks are green! From my side this is good to be merged

davsclaus commented:
LGTM

Thanks for the effort @christophd and sorry for the bike-shedding but thats the nature of humans, github and big PRs ;)

oscerd commented:
Thanks. Nice addition!
